taiga_halo2/examples/sudoku/app/valid_sudoku.rs

```diff
@@ -0,0 +1,447 @@
+            }
+        }
+
+        // check that rows, cols and small squares are permutations of (1..10)
+        for (cpt, perm) in [rows, cols, squares].concat().iter().enumerate() {
+            // lhs = prod_i (i + gamma)
+            // rhs = -prod_i (sigma(i) + gamma)
+            // compute the circuit corresponding to lhs + rhs (should be zero)
+            let mut cell_lhs = assign_free_advice(
+                layouter.namespace(|| "lhs init"),
+                config.advices[0],
+                Value::known(Fp::one()),
+            )
+            .unwrap();
+            let mut cell_rhs = assign_free_advice(
+                layouter.namespace(|| "rhs init"),
+                config.advices[1],
+                Value::known(-Fp::one()),
+            )
+            .unwrap();
+            for i in 1..10 {
+                // LHS
+                // t = i + gamma
+                let t = AddInstructions::add(
+                    &config.add_chip(),
+                    layouter.namespace(|| "i+gamma"),
+                    &cell_integers[i - 1],
+                    &gamma,
+                )
+                .unwrap();
+                // lhs = lhs * t
+                cell_lhs = MulInstructions::mul(
+                    &config.mul_chip(),
+                    layouter.namespace(|| "lhs * t"),
+                    &cell_lhs,
+                    &t,
+                )
+                .unwrap();
+
+                // RHS
+                // u = sigma(i) + gamma
+                let u = AddInstructions::add(
+                    &config.add_chip(),
+                    layouter.namespace(|| "sigma(i)+gamma"),
+                    &perm[i - 1],
+                    &gamma,
+                )
+                .unwrap();
+                // rhs = rhs * u
+                cell_rhs = MulInstructions::mul(
+                    &config.mul_chip(),
+                    layouter.namespace(|| "rhs * u"),
+                    &cell_rhs,
+                    &u,
+                )
+                .unwrap();
+            }
+
+            // We add an instance of LHS + RHS, expected to be zero
+            let expected_zero = AddInstructions::add(
+                &config.add_chip(),
+                layouter.namespace(|| "final add"),
+                &cell_lhs,
+                &cell_rhs,
+            )
+            .unwrap();
+
+            layouter
+                .constrain_instance(expected_zero.cell(), config.primary, cpt)
+                .unwrap();
+        }
+
+        // Check the solution corresponds to the given puzzle
+        sudoku_cells
+            .iter()
+            .enumerate()
+            .for_each(|(i, solution_cell)| {
+                let puzzle_cell = assign_free_instance(
+                    layouter.namespace(|| "instance"),
+                    config.primary,
+                    i + 27,
+                    config.advices[0],
+                )
+                .unwrap();
```

Comment:
Do we need to public the puzzle_cell here? not witness the puzzle_cell

Reply:
The puzzle is a public input

Reply:
sorry, it's already public. I misread `assign_free_instance` as `assign_free_advice`.
You added a new `assign_free_instance`. That's very useful!